src/tof/model.py

```diff
@@ -62,177 +78,187 @@ def run(self, npulses: int = 1):
             sizes=self.pulse.birth_times.sizes, unit=None, dtype=bool
         )
-        for comp in components:
+
+        result_choppers = {}
+        result_detectors = {}
+        for c in components:
+            comp = Component(**c.to_dict())
             comp._wavelengths = self.pulse.wavelengths
             t = self.pulse.birth_times + comp.distance / self.pulse.speeds
             comp._arrival_times = t
-            if isinstance(comp, Detector):
+            if isinstance(c, Detector):
                 comp._mask = initial_mask
+                result_detectors[comp.name] = comp
                 continue
             m = sc.zeros(sizes=t.sizes, unit=None, dtype=bool)
-            to = comp.open_times
-            tc = comp.close_times
+            to = c.open_times
+            tc = c.close_times
             for i in range(len(to)):
                 m |= (t > to[i]) & (t < tc[i])
             combined = initial_mask & m
             comp._mask = combined
             comp._own_mask = ~m & initial_mask
             initial_mask = combined
+            result_choppers[comp.name] = comp
 
-    def _add_rays(self, ax, tofs, birth_times, distances, cbar=True, wavelengths=None):
-        x0 = birth_times.to(unit='us', copy=False).values.reshape(-1, 1)
-        x1 = tofs.to(unit='us', copy=False).values.reshape(-1, 1)
-        y0 = np.zeros(x0.size).reshape(-1, 1)
-        y1 = distances.values.reshape(-1, 1)
-        segments = np.concatenate(
-            (
-                np.concatenate((x0, y0), axis=1).reshape(-1, 1, 2),
-                np.concatenate((x1, y1), axis=1).reshape(-1, 1, 2),
-            ),
-            axis=1,
+        return Result(
+            pulse=self.pulse, choppers=result_choppers, detectors=result_detectors
         )
-        coll = LineCollection(segments)
-        if wavelengths is not None:
-            coll.set_cmap(plt.cm.gist_rainbow_r)
-            coll.set_array(wavelengths.values)
-            coll.set_norm(plt.Normalize(self.pulse.lmin.value, self.pulse.lmax.value))
-            if cbar:
-                cb = plt.colorbar(coll)
-                cb.ax.yaxis.set_label_coords(-0.9, 0.5)
-                cb.set_label('Wavelength (Å)')
-        else:
-            coll.set_color('lightgray')
-        ax.add_collection(coll)
 
-    def plot(
-        self,
-        max_rays: int = 1000,
-        blocked_rays: int = 0,
-        figsize=None,
-        ax=None,
-        cbar=True,
-    ) -> tuple:
-        """
-        Plot the time-distance diagram for the instrument, including the rays of
-        neutrons that make it to the furthest detector.
-        As plotting many lines can be slow, the number of rays to plot can be
-        limited by setting ``max_rays``.
-        In addition, it is possible to also plot the rays that are blocked by
-        choppers along the flight path by setting ``blocked_rays > 0``.
+    # def _add_rays(self, ax, tofs, birth_times, distances, cbar=True, wavelengths=None):
```

Comment:
commented code

Reply:
Yes, please note that I am not at all done with my changes, even though I pushed. I sometimes have to do this when I'm in the middle of something but I have to leave. I will send you a message when I am ready for another review.